File: benchmarks/go/binary_trees.go
// Binary trees benchmark adapted from The Computer Language Benchmarks Game

package main

import (
	"fmt"
	"runtime"
	"time"
)

// A TreeNode structure
type TreeNode struct {
	left  *TreeNode
	right *TreeNode
	item  int
}

// Create a new tree with the given item value at the root
func NewTree(item, depth int) *TreeNode {
	if depth <= 0 {
		return &TreeNode{nil, nil, item}
	}
	return &TreeNode{
		NewTree(2*item-1, depth-1),
		NewTree(2*item, depth-1),
		item,
	}
}

// Check the tree and return a checksum
func CheckTree(node *TreeNode) int {
	if node == nil {
		return 0
	}
	if node.left == nil {
		return node.item
	}
	return node.item + CheckTree(node.left) - CheckTree(node.right)
}

func main() {
	minDepth := 4
	maxDepth := 12

	stretchDepth := maxDepth + 1
	startTime := time.Now()

	// Allocate and check a big tree
	bigTree := NewTree(0, stretchDepth)
	fmt.Printf("stretch tree of depth %d check: %d\n", stretchDepth, CheckTree(bigTree))

	// Allow the big tree to be garbage collected
	bigTree = nil

	// Allocate a long-lived binary tree
	longLivedTree := NewTree(0, maxDepth)

	// Check trees of increasing depth
	for depth := minDepth; depth <= maxDepth; depth += 2 {
		iterations := 1 << (maxDepth - depth + minDepth)
		result := 0

		for i := 0; i < iterations; i++ {
			a := NewTree(i, depth)
			b := NewTree(-i, depth)
			result += CheckTree(a) + CheckTree(b)
		}

		fmt.Printf("%d trees of depth %d check: %d\n", iterations*2, depth, result)
	}

	// Check the long-lived tree last
	fmt.Printf("long lived tree of depth %d check: %d\n", maxDepth, CheckTree(longLivedTree))

	elapsed := time.Since(startTime).Milliseconds()
	fmt.Printf("Time taken: %d ms\n", elapsed)

	// Get memory stats
	var mem runtime.MemStats
	runtime.ReadMemStats(&mem)
	fmt.Printf("Memory used: %d KB\n", mem.Alloc/1024)
}

File: benchmarks/rust/binary_trees.rs
// Binary trees benchmark adapted from The Computer Language Benchmarks Game

use std::rc::Rc;
use std::cell::RefCell;
use std::time::Instant;

// A TreeNode structure
struct TreeNode {
    left: Option<Rc<RefCell<TreeNode>>>,
    right: Option<Rc<RefCell<TreeNode>>>,
    item: i32,
}

impl TreeNode {
    // Create a new tree with the given item value at the root
    fn new(item: i32, depth: i32) -> Rc<RefCell<TreeNode>> {
        let mut node = TreeNode { left: None, right: None, item };
        
        if depth > 0 {
            node.left = Some(TreeNode::new(2 * item - 1, depth - 1));
            node.right = Some(TreeNode::new(2 * item, depth - 1));
        }
        
        Rc::new(RefCell::new(node))
    }
    
    // Check the tree and return a checksum
    fn check(&self) -> i32 {
        match (&self.left, &self.right) {
            (None, None) => self.item,
            (Some(left), Some(right)) => {
                self.item + left.borrow().check() - right.borrow().check()
            },
            _ => unreachable!("Tree should always have either both children or no children"),
        }
    }
}

fn main() {
    let min_depth = 4;
    let max_depth = 12;
    
    let stretch_depth = max_depth + 1;
    let start_time = Instant::now();
    
    // Allocate and check a big tree
    let big_tree = TreeNode::new(0, stretch_depth);
    println!("stretch tree of depth {} check: {}", stretch_depth, big_tree.borrow().check());
    
    // Allow the big tree to be garbage collected
    drop(big_tree);
    
    // Allocate a long-lived binary tree
    let long_lived_tree = TreeNode::new(0, max_depth);
    
    // Check trees of increasing depth
    for depth in (min_depth..=max_depth).step_by(2) {
        let iterations = 1 << (max_depth - depth + min_depth);
        let mut result = 0;
        
        for i in 0..iterations {
            let a = TreeNode::new(i, depth);
            let b = TreeNode::new(-i, depth);
            result += a.borrow().check() + b.borrow().check();
        }
        
        println!("{} trees of depth {} check: {}", iterations*2, depth, result);
    }
    
    // Check the long-lived tree last
    println!("long lived tree of depth {} check: {}", max_depth, long_lived_tree.borrow().check());
    
    let elapsed = start_time.elapsed();
    println!("Time taken: {} ms", elapsed.as_millis());
    
    // Get approximate memory usage - note this isn't as accurate as in other languages
    let memory_usage = std::mem::size_of::<TreeNode>() * (1 << (max_depth + 1)) / 1024;
    println!("Memory used: {} KB", memory_usage);
}